cpp/src/rolling/range_rolling.cu

```diff
@@ -0,0 +1,234 @@
+#include <cudf/rolling/range_window_bounds.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_checks.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cub/device/device_segmented_reduce.cuh>
+#include <thrust/functional.h>
+#include <thrust/reduce.h>
+
+#include <optional>
+
+namespace CUDF_EXPORT cudf {
+namespace detail {
+
+/**
+ * @brief Compute the number of nulls in each group.
+ *
+ * @param orderby Column with null mask.
+ * @param offsets Offset array defining the (sorted) groups.
+ * @param stream CUDA stream used for kernel launches
+ * @return device_uvector containing the null count per group.
+ */
+[[nodiscard]] rmm::device_uvector<cudf::size_type> nulls_per_group(
+  column_view const& orderby,
+  rmm::device_uvector<size_type> const& offsets,
+  rmm::cuda_stream_view stream)
+{
+  auto d_orderby        = column_device_view::create(orderby, stream);
+  auto const num_groups = offsets.size() - 1;
+  std::size_t bytes{0};
+  auto is_null_it = cudf::detail::make_counting_transform_iterator(
+    cudf::size_type{0}, [orderby = *d_orderby] __device__(size_type i) -> size_type {
+      return static_cast<size_type>(orderby.is_null_nocheck(i));
+    });
+  rmm::device_uvector<cudf::size_type> null_counts{num_groups, stream};
+  cub::DeviceSegmentedReduce::Sum(nullptr,
+                                  bytes,
+                                  is_null_it,
+                                  null_counts.begin(),
+                                  num_groups,
+                                  offsets.begin(),
+                                  offsets.begin() + 1,
+                                  stream.value());
+  auto tmp = rmm::device_buffer(bytes, stream);
+  cub::DeviceSegmentedReduce::Sum(tmp.data(),
+                                  bytes,
+                                  is_null_it,
+                                  null_counts.begin(),
+                                  num_groups,
+                                  offsets.begin(),
+                                  offsets.begin() + 1,
+                                  stream.value());
+  return null_counts;
+}
+
+/**
+ * @brief Deduce the `null_order` of a column given sort order and group offsets
+ *
+ * @param orderby The orderby column to check.
+ * @param order The sort order of the column.
+ * @param offsets Group offsets.
+ * @param stream CUDA stream used for kernel launches.
+ *
+ * @return The deduced `null_order`.
+ */
+[[nodiscard]] null_order deduce_null_order(column_view const& orderby,
+                                           order order,
+                                           rmm::device_uvector<size_type> const& offsets,
+                                           rmm::device_uvector<size_type> const& per_group_nulls,
+                                           rmm::cuda_stream_view stream)
```

Comment:
This is needed to polyfill the old interface to the new one.

Reply:
Ah, thank you for this.